src/text_shaper/directwrite_shaper.cpp

```diff
@@ -279,58 +284,137 @@ void directwrite_shaper::shape(font_key _font,
-    // auto glyphProps = new DWRITE_SHAPING_GLYPH_PROPERTIES[maxGlyphCount];
-    // auto actualGlyphCount = UINT32{0};
-
-    // analyzer->GetGlyphs(textString,
-    //                     textLength,
-    //                     fontFace,
-    //                     isSideways,
-    //                     isRightToLeft,
-    //                     scriptAnalysis,
-    //                     localeName, numberSubstitution,
-    //                     &features/*FIXME: is this right?*/,
-    //                     featureRangeLengths,
-    //                     featureRanges,
-    //                     maxGlyphCount,
-    //                     clusterMap,
-    //                     textProps,
-    //                     glyphIndices,
-    //                     glyphProps,
-    //                     &actualGlyphCount);
-
+    std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> conv1;
+    std::string bytes = conv1.to_bytes(std::u32string{ _text });
+    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> conv2;
+    std::wstring wText = conv2.from_bytes(bytes);
+
+    WCHAR const* textString = wText.c_str();
+    UINT32 textLength = wText.size();
+    FontInfo const& fontInfo = d->fonts.at(_font);
+    IDWriteFontFace5* fontFace = fontInfo.fontFace.Get();
+
+    std::vector<UINT16> _glyphIndices;
+    std::vector<INT32> _glyphDesignUnitAdvances;
+
+    BOOL isTextSimple = FALSE;
+    UINT32 uiLengthRead = 0;
+    const UINT32 glyphStart = 0;
+    _glyphIndices.resize(textLength);
+
+    d->textAnalyzer->GetTextComplexity(
+        textString,
+        textLength,
+        fontFace,
+        &isTextSimple,
+        &uiLengthRead,
+        &_glyphIndices.at(glyphStart));
+
+    BOOL _isEntireTextSimple = isTextSimple && uiLengthRead == textLength;
+    if (_isEntireTextSimple)
+    {
+        DWRITE_FONT_METRICS1 metrics;
+        fontFace->GetMetrics(&metrics);
+
+        _glyphDesignUnitAdvances.resize(textLength);
+        USHORT designUnitsPerEm = metrics.designUnitsPerEm;
+        fontFace->GetDesignGlyphAdvances(
+            textLength,
+            & _glyphIndices.at(glyphStart),
+            & _glyphDesignUnitAdvances.at(glyphStart),
+            false);
+
+        for (size_t i = glyphStart; i < textLength; i++)
+        {
+            const auto cellWidth = static_cast<double>((float)_glyphDesignUnitAdvances.at(i)) / designUnitsPerEm * fontInfo.size.pt;
+            glyph_position gpos{};
+            gpos.glyph = glyph_key{ _font, fontInfo.size, glyph_index{_glyphIndices.at(i)} };
+            gpos.offset.x = static_cast<int>(cellWidth);
+            //gpos.offset.y = static_cast<int>(static_cast<double>(pos[i].y_offset) / 64.0f);
+
+            gpos.advance.x = static_cast<int>(cellWidth);
+            //gpos.advance.y = static_cast<int>(static_cast<double>(pos[i].y_advance) / 64.0f);
+            _result.emplace_back(gpos);
+        }
+    }
+    else {
+        // TODO
```

Comment:
This is a BIG TODO here. No complex shaping is supported right now. Consolas will report most ASCII text as "simple", but other fonts might not do so. Hard-code `_isEntireTextSimple` if you need it for testing.

Reply:
When this function is called, the text is already segmented by the font properties needed to do proper text shaping, i.e.

* the text sequence contains only one script (e.g. Latin, or Greek but never both)
* either emoji emoji presentation **or** emoji text presentation
* split up by words (space-delimited)

Reply:
OK hard-code `_isEntireTextSimple ` won't work because `_glyphIndices` isn't filled with the correct indices...Just use Consolas for testing 😅